### PR TITLE
Transfer iterator

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

File: inc/transfer.class.php [hunk 1]
```diff
@@ -718,145 +766,212 @@ function simulateTransfer($items) {
...
+            'FROM'      => 'glpi_tickets',
+            'LEFT JOIN' => [
+               'glpi_suppliers_tickets'   => [
+                  'ON' => [
+                     'glpi_suppliers_tickets'   => 'tickets_id',
+                     'glpi_tickets'             => 'id'
+                  ]
+               ],
+               'glpi_suppliers'           => [
+                  'FKEY' => [
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/transfer.class.php]: :D

File: inc/transfer.class.php [hunk 2]
```diff
@@ -277,10 +276,7 @@ function simulateTransfer($items) {
       global $DB, $CFG_GLPI;
 
       // Init types :
-      $types = ['CartridgeItem', 'Change', 'Computer', 'ConsumableItem', 'Contact', 'Contract',
-                     'Document', 'Link', 'Monitor', 'NetworkEquipment', 'Peripheral', 'Phone',
-                     'Printer', 'Problem', 'Software', 'SoftwareLicense', 'SoftwareVersion',
-                     'Supplier', 'Ticket'];
+      $types = $this->inventory_types;
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/transfer.class.php]: @orthagh any opinion?

[the remark this answers is not in the document]
Comment (orthagh) [reply inc/transfer.class.php]: indeed, this list differs, i think it represents the linked items of an inventory type, not also himself an inventory type

File: inc/transfer.class.php [hunk 3]
```diff
@@ -908,109 +1033,126 @@ function simulateTransfer($items) {
...
-            = $this->createSearchConditionUsingArray($this->noneedtobe_transfer['Document']);
-
       // printer -> cartridges : keep / delete + clean
       if ($this->options['keep_cartridgeitem']) {
-         if (isset($this->item_search['Printer'])) {
-            $query = "SELECT `cartridgeitems_id`
-                      FROM `glpi_cartridges`
-                      WHERE `printers_id` IN ".$this->item_search['Printer'];
+         if (isset($this->needtobe_transfer['Printer'])) {
+            $iterator = $Db->requests([
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
            $iterator = $DB->request([
```

File: inc/transfer.class.php [hunk 4]
```diff
@@ -1921,153 +2067,158 @@ function transferDocuments($itemtype, $ID, $newID) {
...
+                  );
+               }
+
             }
-         }
 
+            // If clean and unused ->
+            if ($need_clean_process
+                  && $this->options['clean_document']) {
+               $remain = $Db->request([
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
               $remain = $DB->request([
```

File: inc/transfer.class.php [hunk 5]
```diff
@@ -2121,160 +2272,167 @@ function transferDirectConnection($itemtype, $ID, $link_type) {
...
-                           if ($newID != $item_ID) {
-                              $need_clean_process = true;
-                           }
+         $criteria['WHERE']['NOT'] = ['items_id' => $this->noneedtobe_transfer[$link_type]];
+      }
+
+      $iterator = $DB->request($criteria);
+
+      // Foreach get item
+      while ($data = $DB->fetch_assoc($result)) {
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
      while ($data = $iterator->next()) {
```

File: inc/transfer.class.php [hunk 6]
```diff
@@ -2439,50 +2594,50 @@ function transferLinkedSuppliers($itemtype, $ID, $newID) {
...
-      if ($result = $DB->query($query)) {
-         if ($DB->numrows($result) != 0) {
-            while ($data = $DB->fetch_assoc($result)) {
-               $input = [];
-
-               if ($data['suppliers_id'] > 0) {
-                  $supplier = new Supplier();
+      $iterator = $DB->request([
+         'FROM'   => $table,
+         'WHERE'  => [$field => $id]
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
         'WHERE'  => [$field => $ID]
```

File: inc/transfer.class.php [hunk 7]
```diff
@@ -2520,37 +2675,32 @@ function transferTaskCategory($itemtype, $ID, $newID) {
...
-                FROM `$table`
-                WHERE `$field` = '$ID'";
-
-      if ($result = $DB->query($query)) {
-         if ($DB->numrows($result) != 0) {
-            while ($data = $DB->fetch_assoc($result)) {
-               $input = [];
+      $iterator = $DB->request([
+         'FROM'   => $table,
+         'WHERE'  => [$field => $id]
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
         'WHERE'  => [$field => $ID]
```

File: inc/transfer.class.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (trasher) [reply inc/transfer.class.php]: ok, I'll revert this part :)

File: inc/transfer.class.php [hunk 8]
```diff
@@ -2655,20 +2805,19 @@ function transferCompatiblePrinters($ID, $newID) {
...
+         $iterator = $DB->request([
+            'FROM'   => 'glpi_cartridgeitems_printermodels',
+            'WHERE'  => ['cartridgeitems_id' => $ID]
+         ]);
 
-               while ($data = $DB->fetch_assoc($result)) {
-                  $data = Toolbox::addslashes_deep($data);
-                  $cartitem->addCompatibleType($newID, $data["printermodels_id"]);
-               }
+         if ($count($iterator)) {
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
         if (count($iterator)) {
```

File: inc/transfer.class.php [hunk 9]
```diff
@@ -2842,138 +3001,139 @@ function transferSupplierContacts($ID, $newID) {
...
-                                                    NOT IN ".$this->item_search['Supplier'] ."
-                                               AND `suppliers_id`
-                                                    NOT IN ".$this->item_recurs['Supplier'];
-                        $result_search = $DB->query($query_search);
-                        if ($DB->result($result_search, 0, 'cpt') > 0) {
-                           $canbetransfer = false;
-                        }
-                     }
+         // Foreach get item
+         while ($data = $DB->fetch_assoc($result)) {
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
         while ($data = $iterator->next()) {
```

File: inc/transfer.class.php [hunk 10]
```diff
@@ -2842,138 +3001,139 @@ function transferSupplierContacts($ID, $newID) {
...
+            // If clean and unused ->
+            if ($need_clean_process
+                  && $this->options['clean_contact']) {
+               $remain = $DB->request([
+                  'COUNT'  => 'cpt',
+                  'FROM'   => 'glpi_contacts_suppliers',
+                  'WHERE'  => ['contacts_id' => $item_ID]
+               ])->next();
+
+               if ($result['cpt'] == 0) {
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
               if ($remain['cpt'] == 0) {
```

File: inc/transfer.class.php [hunk 11]
```diff
@@ -3057,116 +3217,125 @@ function transferDevices($itemtype, $ID, $newID) {
...
+                        // No
+                        // Can be transfer without copy ? = all linked items need to be transfer (so not copy)
+                        $canbetransfer = true;
+                        $type_iterator = $DB->request([
+                           'SELECT DISTINCT' => 'itemtype',
+                           'FROM'            => $itemdevicetable,
+                           'WHERE'           => [$fk => $item_ID]
+                        ]);
+
+                        while (($data_type = $iterator->next())
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
                        while (($data_type = $type_iterator->next())
```

File: inc/transfer.class.php [hunk 12]
```diff
@@ -3188,81 +3357,92 @@ function transferNetworkLink($itemtype, $ID, $newID) {
...
+                  while ($data = $iterator->next()) {
+                     unset($data['id']);
+                     $data['items_id'] = $newID;
+                     $data['netpoints_id']
+                                       = $this->transferDropdownNetpoint($data['netpoints_id']);
+                     unset($np->fields);
+                     $np->add(toolbox::addslashes_deep($data));
+                  }
+               } else {
+                  while ($data = $DB->fetch_assoc($result)) {
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
                  while ($data = $iterator->next()) {
```

File: inc/transfer.class.php [hunk 13]
```diff
@@ -376,104 +376,131 @@ function simulateTransfer($items) {
...
-                     $this->addNotToBeTransfer('SoftwareVersion', $data['vID']);
-                  } else {
-                     $this->addToBeTransfer('SoftwareVersion', $data['vID']);
-                  }
+         $iterator = $DB->request([
+            'SELECT'       => [
+               'glpi_softwares.id',
+               'glpi_softwares.entities_id',
+               'glpi_softwares.is_recursive',
+               'glpi_softwareversions`.`id` AS vID'
```
Comment (cedric-anne) [inline inc/transfer.class.php]:
```suggestion
               'glpi_softwareversions.id` AS vID'
```

**Pull request merged**